### PR TITLE
use async-channel to notify senders of disconnection

### LINKED ISSUE #4: Reevaluate the Arc<Mutex> used to track disconnection
this issue exists as a bookmark that there might be improvements over using an `Arc<Mutex>`, such as possibly something from crossbeam

### LINKED ISSUE #5: Use async-channel
https://github.com/http-rs/async-sse/pull/3 adds support for detecting disconnections through a custom `Arc<AtomicBool>`. This is not particularly efficient, but it works.

https://docs.rs/async-channel/1.1.1/async_channel/'s channels have a Sender that returns an error on disconnect; exactly as we would expect here. We should use this instead of our custom `AtomicBool` impl. `async-std` is expected to move over to this as well in the near future, but we shouldn't wait for that.

@jbr could you pick this up?

### PR DESCRIPTION
**Pull request by jbr**
This reverts #3 and uses async-channel instead
closes #4 
closes #5 

I realized after pushing this that it probably would have been easier to read as two distinct commits. This is probably the clearest view of the changes: https://github.com/http-rs/async-sse/compare/886cd7551048d119dbad011fd9b31b56a07eddb8...95ddc4bf6862dec5f597d487104bf7826f79836d

### CODE REVIEW

File: tests/encode.rs [hunk 1]
```diff
@@ -65,16 +65,17 @@ async fn encode_retry() -> http_types::Result<()> {
...
+
+    std::mem::drop(reader);
 
-    assert_eq!(
-        sender.send("cat", "nori", None).await,
-        Err(async_sse::DisconnectedError)
-    );
+    assert!(matches!(sender.send("cat", "chashu", None).await,
+                     Err(e @ async_std::io::Error { .. })
+                            if e.kind() == async_std::io::ErrorKind::ConnectionAborted));
```
Comment (yoshuawuyts) [inline tests/encode.rs:79]: This assert is a bit hard to follow.

File: src/encoder.rs [hunk 2]
```diff
@@ -90,80 +78,63 @@ impl AsyncRead for Encoder {
...
         let msg = format!("retry:{}\n\n", dur);
-        self.sender.send(msg.into_bytes()).await;
+        self.inner_send(msg).await?;
+        Ok(())
+    }
+}
+
+impl Clone for Sender {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
```
Comment (yoshuawuyts) [inline src/encoder.rs:138]: Nice!

Comment (jbr) [reply src/encoder.rs:138]: This bit didn't actually change, although probably it should have. In the intermediate commit I changed this to `#[derive(Clone)]` because it's identical, but that got reverted

File: tests/encode.rs (hunk 1 above)
Comment (jbr) [reply tests/encode.rs:79]: Agreed, will clean it up

**Comment by jbr** [comment]
Although I believe these changes are still worthwhile because async-sse should report message send failure, I think I just found a way to address the connection-closed issue more generally in tide itself

File: tests/encode.rs (hunk 1 above)
Comment (jbr) [reply tests/encode.rs]: addressed in 95ddc4bf6862dec5f597d487104bf7826f79836d

File: src/encoder.rs (hunk 2 above)
Comment (jbr) [reply src/encoder.rs]: switched back in f484fd446cb62f9b643e18ef49ad9b5403fc9872, lmk if there's a reason _not_ to use the derived clone impl

Comment (yoshuawuyts) [reply src/encoder.rs]: @jbr :100:
